### DISCUSSION TITLE
Why do exception handlers expect Exceptions and not BaseExceptions

### DISCUSSION THREAD

**Discussion by sidekick-martin** [fastapi/fastapi#9379] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from starlette.requests import Request
from starlette.responses import JSONResponse

def base_exception(request: Request, exception: BaseException) -> JSONResponse:
    return JSONResponse(content=request.json())

app = FastAPI(exception_handlers={BaseException: base_exception})
```

### Description

Run mypy check on the example code.

I get:
```example.py:10: error: Dict entry 0 has incompatible type "Type[BaseException]": "Callable[[Request, BaseException], JSONResponse]"; expected "Union[int, Type[Exception]]": "Callable[[Request, Any], Coroutine[Any, Any, Response]]"  [dict-item]```

Is there a reason for using `Exception` specifically for typing here instead of `BaseException`?

### Operating System

macOS

### FastAPI Version

0.95.0

### Python Version

3.10.7

**Comment by n8sty** [comment, marked as answer]
`BaseException` is the base class that all built-in exceptions derive from (I'm paraphrasing / plagiarizing the docs here https://docs.python.org/3/library/exceptions.html#BaseException) and should not be used as base for any user defined errors. This means for all intents and purposes that `BaseException` is not something that should be referenced like you're doing and instead should be `Exception`. In effect, `BaseException` is too low-level a type.

**Reply by Torxed** [reply]
Wouldn't it mean tho that we would be able to add more exception handlers to FastAPI via `.exception_handler()` such as built in ones deriving from `BaseException`, if we checked if the exception is instance of `BaseException`?

tl;dr: Are we missing opportunities to catch exceptions (builtin or custom) in FastAPI because we limit ourselves to `Exception`?

**Reply by n8sty** [reply]
Technically opportunities to catch exception are being missed, in reality that shouldn't be an issue. See https://docs.python.org/3/library/exceptions.html#exception-hierarchy for a representation of how Python's inheritance of exceptions works. For example, `KeyboardInterrupt`s won't be caught if you just use `Exception` but... that's a good thing for almost every case I can imagine.

**Reply by Torxed** [reply]
> that's a good thing for almost every case I can imagine.

I agree that those few might not be worth even considering.
I somehow remember the exception stack being more outside of `Exception`. Thanks for the hierarchy link!
